File: backend/src/main/java/com/kiwi/features/metrics/MetricsFactory.java
package com.kiwi.features.metrics;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

import static com.kiwi.utils.FormatUtils.formatDate;

public class MetricsFactory {
    private static final Random RANDOM = new Random();

    public static MetricsDTO generateRandomValidMetricDTO() {
        return new MetricsDTO(getRandomDate(), getRandomSteps(), getRandomScreenTimeSeconds());
    }

    public static MetricsDTO generateRandomInvalidMetricDTO() {
        return new MetricsDTO(getRandomDate(), -getRandomSteps(), -getRandomScreenTimeSeconds());
    }
    
    public static Set<MetricsDTO> generateRandomMetricsSet(int size, boolean getValidValues) {
        Set<MetricsDTO> metricsDTOSet = new HashSet<>();
        for (int i = 0; i < size; i++) {
            metricsDTOSet.add(getValidValues ? generateRandomValidMetricDTO() : generateRandomInvalidMetricDTO());
        }
        return metricsDTOSet;
    }
    
    private static String getRandomDate() {
        int startYear = 2025;
        int endYear = 2026;

        int year = RANDOM.nextInt(endYear - startYear + 1) + startYear;
        int month = RANDOM.nextInt(12) + 1;
        int dayOfMonth = RANDOM.nextInt(28) + 1;

        return formatDate(LocalDate.of(year, month, dayOfMonth));
    }
    
    private static Integer getRandomSteps() {
        return RANDOM.nextInt(10001);
    }
    
    private static Integer getRandomScreenTimeSeconds() {
        return RANDOM.nextInt(4 * 60 * 60) + 60;
    }
}

File: mobile/app/src/main/java/com/bellako/kiwi/features/metrics/MetricsFactory.kt
package com.bellako.kiwi.features.metrics

import android.os.Build
import androidx.annotation.RequiresApi
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import kotlin.random.Random

@RequiresApi(Build.VERSION_CODES.O)
object MetricsFactory {
    private val formatter = DateTimeFormatter.ISO_LOCAL_DATE

    fun generateRandomValidMetricDTO(): MetricsDTO {
        return MetricsDTO(
            email = "finn@thehuman.com",
            date = getRandomDate(),
            steps = getRandomSteps(),
            screenTimeSeconds = getRandomScreenTimeSeconds()
        )
    }

    fun generateRandomInvalidMetricDTO(): MetricsDTO {
        return MetricsDTO(
            email = "finn@thehuman.com",
            date = getRandomDate(),
            steps = -getRandomSteps(),
            screenTimeSeconds = -getRandomScreenTimeSeconds()
        )
    }

    fun generateRandomMetricsSet(size: Int, valid: Boolean): Set<MetricsDTO> {
        return (1..size).map {
            if (valid) generateRandomValidMetricDTO() else generateRandomInvalidMetricDTO()
        }.toSet()
    }

    private fun getRandomDate(): String {
        val year = Random.nextInt(2025, 2027)
        val month = Random.nextInt(1, 13)
        val day = Random.nextInt(1, 29)

        return LocalDate.of(year, month, day).format(formatter)
    }

    private fun getRandomSteps(): Int {
        return Random.nextInt(1, 10001)
    }

    private fun getRandomScreenTimeSeconds(): Int {
        return Random.nextInt(60, 4 * 60 * 60 + 1)
    }
}